README.md


# 
SUAI CLI TIMETABLE

A small CLI app written in Python3 for my own use and 4 educational purposes.
Grasp provides information about the schedule of classes in my university (SUAI). The information is parsed from the official site (rasp.guap.ru) and is provided in a convenient form, with all possible selections, filters and other, the parity of the week is also taken into account. For example - schedule for tomorrow, schedule for Tuesday next week and so on - is a correct request for grasp. A feature is the possibility of correct operation in offline. In offline mode, the current day of the week, the parity of the week and so on are correctly calculated.
Caching takes place either on a separate key or at the end of the request for a timetable on-line in order to remind the user to periodically re-cache the schedule


## Getting Started

Currently worling only under python 3.x. The only thing u need 2 run this util is Python! Just download it from python.org

### Installing

After installing python u re ready to install the app :)

**pip / Github (THE EASIEST WAY)**
```
pip install git+https://github.com/maxan98/grasp.git
or
pip3 install git+https://github.com/maxan98/grasp.git
```

or

```
git clone https://github.com/maxan98/grasp.git
python grasp/setup.py install
```
**Just download (Like the way it used to be)**
```
wget -O grasp https://raw.githubusercontent.com/maxan98/grasp/master/grasp.py
chmod +x grasp
```
or

```
curl -Lo grasp https://raw.githubusercontent.com/maxan98/grasp/master/grasp.py
chmod +x grasp
```

End with an example of getting some data out of the system or using it for a little demo

## Usage/help/screenshots/additional info

Auto upgrades when new version is available.
If u want 2 force-upgrade the app use
```
pip3 install -U git+https://github.com/maxan98/grasp.git
```

All info and help u can find just running with -h key.
Output of -h key:
![2018-03-07 14 57 08](https://user-images.githubusercontent.com/24477803/37091064-d99522f0-2217-11e8-97bb-e1efcaa7f266.png)



Supports offline as well as online mod.

2 use offline mod, u ned to cache timetable for your group first:

`grasp -c(--cache) -g(--group) xxxxx(5512 is default)`

or just check your Timetable once in online mod. After that u will be asked 2 cache your tt automatically:

`Would you like to cache timetable for this group? [y/N]`

Jyst type **YES**!

After that u can use offline mod just running with **-f** key

` grasp -f -g 8431К -d mon`

U can use variaty of different usefull keys.

Also, u r able 2 combine them like this:

` grasp -ftg 8431К` 

With command above u call 4 --offline --today and --group :) As simple as possible)

There r only 3 REQUIRED keys. Choose one of them or app wont run.
-o --online

-f --offline

-c --cache

All other keys can be omitted.

If omitted:

Default for group is 5512.
For day is WHOLE
For week is both

**!!!WARNING!!! Notice that if your group numbers contains letters - they r in cyrillic. After decoding utf-8 cyrillic letters and latin letters are not the same! For ex. '8431К'. 'K' must be entered in cyrillic**

Offline mod r very useful cuz SUAI adores shutting down their servers and don't turn the half-week

Im still working on some features and on prettyfying the output.

**Now it looks like this :)**

![2018-03-04 21 16 27](https://user-images.githubusercontent.com/24477803/36948864-e09ae29c-1ff1-11e8-8194-b5a18d942ae8.png)

###### PREVIOUS VERSIONs SCREENSHOTS

![2018-03-04 1 04 46](https://user-images.githubusercontent.com/24477803/36939788-42c074f0-1f48-11e8-9a3f-fc0210a50cd0.png)


changelog.txt
----------TODO-----------
[-]Возможность добавлять кастомные хуевины(пары которые на словах препод установил и тд) в расписание
[+ удалил util нахер через pip ставим]СЛЕТАЕТ СИМЛИНК - ХЗ ЧЕ ЗА ШЛЯПА - перстает определять текущую неделю и соответственно в оффлайне файлы не находит]
Наткнулся на БАГ ВЫШЕ и не доделал nweek текущая/красная синяя недели
[+][NWEEK] доделал. 1 баг остался - норм удаляем время перед треугольником но иногда последней пары время не удаляется

[+] ТО ЧТО В ОФФЕ НЕ СЧИТАЕМ ЧЕТНОСТИ НЕДЕЛЬ УЖЕ К БАГАМ ПРИВОДИТ при -t и mon
[+]-оффлайн подсчет четности недели 
[-]Расскрасить аудитории и мб тип занятий (Л, ЛР КР) если имеет смысл
[+]Доделать ключ -w и выдавать расписание синий/красной/текущей/обеих недель в зависимости от ключа
[-]Выборку за несколько дней Ex. -d mtw (monday tuesday wednesday)
[+/-]Расписание на день типа "завтра/послезавтра/через 4 дня"
[-]Предупреждение о праздниках и прочей дрисни
[+]Предупреждение о необходимости с некой периодичностью перекешировать расписание(мб сделать в бэкграунде без спроса юзера вообще хз) - сделал лучше - сверяем по номеру сборки
[-]lxml’s HTML parser	BeautifulSoup(markup, "lxml")	
Very fast
Lenient


--------CHANGELOG--------
init.
04.03.18
[FIXED] определение текущей четности недели
[FIXED] При наличии пар 'вне сетки расписания' - они дублировались при выводе
[FIXED] Неправильно считало количество учебных дней при наличии выходных посреди недели
[DONE]  Раскрасили  инфу о текущей неделе, дни недели в сетке расписания, номера пар и треугольники в зависимости от течности недели
[FIXED] дикое количество минор багов с выводом кусков херово распаршенного текста. Улучшил вывод впринципе
[FIXED] Корректно обрабатываем 'вне сетки расписания'

06.03.18
[FIXED] дня(ей) нормально пересчитать и выводить
[+] -t (--today) отдельный ключ true/false - расписание на сегодня // по идее работает и с оффлайном но осторожно. Плюс пока не настроил вывод по четности текущей недели. То есть выдает на понедельник к примеру но не красную/синюю 
[+] -d (--day) DZ принимает параметр tom, который выдает расписание на завтра. 
[+] Закодированные расписания нужно периодически обновлять, добавил в конец при запросе любого расписания в онлайн моде это    -    Would you like to cache timetable for this group? [y/N]
[minorbugfixes]

07.03.18
[FIXED] Какой то стремный баг: обычно длинна группы - 4 цифры. Однако, бывает говно вроде 8431К. Когда разбирал вывод отсекал группу по кол-ву символов, ибо она отдельно приходила и было нормально. Однако len('8431K') == 5, и len('Среда') тоже == 5 ну и в общем при условии что в СРЕДУ у группы с ДЛИННОЙ == 5 ПЕРВОЙ парой что то ЕСТЬ - все выводилось корректно, НО не выводился номер пары и ее время.
[FIXED] Removed abs paths
[+/- FIXED] Оффлайн подсчет четности недели добавил но чет я хз, сложно тестить и особо не тестил, должно работать и вроде работает но хезе :D Над норм потестить)

11.03.18

[+] Оффлайн подсчет сделан полностью
[+] -w фильтрация по дню неделю сделана полностью
[+] -d Tom -t и тд выдают теперь расписания фильтрую еще и по недели
[+] ';' был в списке расписания и фильтроовался на выводе - приводило к багам. Теперь реально удаляем его.
[+] Не удалялось последнее время пары при условии фильтра по недели если этой пары нет
[+] week() в онлайне при некоторых условиях полную побень считал. исправил.
[+] ПО ОСНОВНОМУ ФУНКЦИОНАЛУ ВСЕ! И НА ДАННЫЙ МОМЕНТ БАГОВ НЕ НАЙДЕНО. ДАЛЕЕ ИСПРАВЛЯЮ МЕЛКИЕ БАГИ ЕСЛИ НАЙДУТСЯ И СМОЛ ФИЧИ В ФУНКЦИОНАЛ
[+] А не, один баг нашел и исправил. При наличии 2-х и более АБСОЛЮТНО одинаковых пар - одно название одна и даже аудитории препод и тд в один день. Все путалось. Ща норм

15.03.18 

[+] Добавил параметр 'c' к '-w' - ТЕКУЩАЯ неделя.
[+] Запрос о кеширования расписания теперь задаем только тогда, когда номер сборки онлайн и оффлайн расписания не совпадает.
[+] Добавил трай кетчи, что совсем с печальными эррорами не вылетала

20.03.18

[+] установка чеерз pip
[+] Shipping updates via github+pip
[+] При запросе расписания в онлайне и при наличии новой версии спрашиваем об обновлении/апдейт в 1 кнопку
[+] Для чека верси -v (local version), -ov (repo's version)
[+] -v verbose debug info
[+] Optimized logic. x3 faster


grasp.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import sys
import os
import re
import time
from bs4 import BeautifulSoup, Tag, NavigableString
import requests
import argparse
from pprint import pprint
from colorama import Fore
from colorama import Style
from colorama import init
import time
import pickle
import subprocess
from tkinter import *
import tkinter.ttk as ttk


debug = False
d = {'mon':[],'tue':[],'wed':[],'thu':[],'fri':[],'sat':[],'sun':[]}
onlinesbor = '-1'
oflinesbor = '-2'

def customizesimmilar(day):
  if debug:
      t = time.time()

  tmp_elem = ''
  for elem in range(len(d[day])):
    tmp_elem = d[day][elem]
    for i in range(elem,len(d[day])):
      if d[day][i] == tmp_elem:
        d[day][i] = d[day][i] + ' '
  if debug:
      print('customizesimmilar',float(time.time()-t))

def customizetimetabletomatchcurrentweek(mode, day):
  if debug:
    t1 = time.time()
  supported_days = ('mon','tue','wed','thu','fri','sat','sun','whole')
  firstp = -1
  secp = -1
  todel = []
  todel1 = []
  atleastonepair = False
  apdel = []
  modechar = ''
  if mode == 1:
    modechar = '▼'
  elif mode == 0:
    modechar = '▲'
  if day == 'whole':
    customizetimetabletomatchcurrentweek(mode,'mon')
    customizetimetabletomatchcurrentweek(mode,'tue')
    customizetimetabletomatchcurrentweek(mode,'wed')
    customizetimetabletomatchcurrentweek(mode,'thu')
    customizetimetabletomatchcurrentweek(mode,'fri')
    customizetimetabletomatchcurrentweek(mode,'sat')
    customizetimetabletomatchcurrentweek(mode,'sun')
    return

  #33 если день передали говяный - шлеп в пиздц
  if day not in supported_days:
    raise ValueError("invalid day:"+day)
    ### При выводе фильтровал ; а они были отдельным элементом и индексы сбвали - был костыль. Ниже блок их УДАЛЯЕТ а не просто маскирует
  for elem in d[day]:
    if ';' in elem:
      apdel.append(d[day].index(elem))
  for i in range(len(apdel)):
    d[day].pop(apdel[i])
    for j in range(i,len(apdel)):
      apdel[j] = apdel[j]-1
  if mode == 2:
    return
  #ОСТАВИТЬ КРАСНУЮ(ВЕРХНЮЮ)
  for elem in d[day]:
    if modechar in elem:
      todel.append(d[day].index(elem))

  ### удаляем время пар когда нужно
  finaldell = []
  for i in range(len(todel)):
    if todel[i]+1 >= len(d[day]):
      break
    if 'пара ' in d[day][todel[i]-1] and 'пара ' in d[day][todel[i]+1]:
      finaldell.append(d[day].index(d[day][todel[i]-1]))

  finaldell.extend(todel)
  finaldell.sort()
  #print(finaldell)
  for i in range(len(finaldell)):
    d[day].pop(finaldell[i])
    for j in range(i,len(finaldell)):
      finaldell[j] = finaldell[j]-1

  if 'пара ' in d[day][len(d[day])-1]:
    d[day].pop()
  if debug:
      print('customizesimmilartomatchcurrentweek',float(time.time()-t1))
    
          


def query_yes_no(question, default="no"):
    """Ask a yes/no question via raw_input() and return their answer.

    "question" is a string that is presented to the user.
    "default" is the presumed answer if the user just hits <Enter>.
        It must be "yes" (the default), "no" or None (meaning
        an answer is required of the user).

    The "answer" return value is True for "yes" or False for "no".
    """
    if debug:
      t2 = time.time()
    valid = {"yes": True, "y": True, "ye": True,
             "no": False, "n": False}
    if default is None:
        prompt = " [y/n] "
    elif default == "yes":
        prompt = " [Y/n] "
    elif default == "no":
        prompt = " [y/N] "
    else:
        raise ValueError("invalid default answer: '%s'" % default)

    while True:
        sys.stdout.write(question + prompt)
        choice = input().lower()
        if default is not None and choice == '':
            return valid[default]
        elif choice in valid:
            return valid[choice]
        else:
            sys.stdout.write("Please respond with 'yes' or 'no' "
                             "(or 'y' or 'n').\n")

    if debug:
        print('queryyesno',time.time()-t2)
def week(r):

  ithinkits_res = -1
  sp = BeautifulSoup(r,'html.parser')
  select = sp.select('em')
  #print(select)
  if select[0].text[0] == '▲':
    ithinkits_res = 1
  elif select[0].text[0] == '▼':
    ithinkits_res = 0
  else:
    ithinkits_res = -1
  if ns.offline:
    with open (os.path.expanduser('~/cached/'+group+'.pkl'),'rb') as pc:
        data = pickle.load(pc)
        t = time.time()
        kek = time.localtime(t)
        if (int(time.strftime('%W',kek)) - int(time.strftime('%W',data)))%2 == 0:
          return ithinkits_res
        else:
          if ithinkits_res == 1:
            return 0
          elif ithinkits_res == 0:
            return 1
          else:
            return -1

  else:
    return ithinkits_res

def prpr(d):
    for i in d:
      if i.count(';') != 0:
        if isinstance(i, NavigableString):
          continue
        i.remove(';')
      print(i,sep='\n')
    



def site():
    if debug:
      t4 = time.time()
    r = requests.get("http://rasp.guap.ru/").content.decode('utf-8')
    soup = BeautifulSoup(r, "html.parser")
    select = soup.find('option',text=group)
    group_prefix = select.attrs['value']

    r = requests.get("http://rasp.guap.ru/?g="+group_prefix).content.decode('utf-8')
    if debug:
      print('site',time.time()-t4)
    return r
def parseonline(r):

    # r = requests.get("http://rasp.guap.ru/").content.decode('utf-8')
    # soup = BeautifulSoup(r, "html.parser")
    # select = soup.find('option',text=group)
    # group_prefix = select.attrs['value']

    # r = requests.get("http://rasp.guap.ru/?g="+group_prefix).content.decode('utf-8')
    if debug:
      t5 = time.time()
    if week(r) == 0:
      print(Fore.BLUE+Style.BRIGHT+'Сейчас ▼ неделя'+Style.RESET_ALL)
    elif week(r) == 1:
      print(Fore.RED+Style.BRIGHT+'Сейчас ▲ неделя'+Style.RESET_ALL)
    else:
      print('У меня траблы с определением недели')
    Soup = BeautifulSoup(r, 'html.parser')
    days = list(Soup.select('h3'))

    day = []
    #print(day)
    for dy in days:
        day.append(dy.text)

    daysii = list(Soup.select('h4'))

    dayii = []
    #print(dayii)
    for dy in daysii:
        dayii.append(dy.text)

    lenofday = len(day)
    for i in day:
      if i == 'Вне сетки расписания':
        lenofday -= 1
    if lenofday == 1:
      print("Учимся всего", lenofday, "день")
    elif lenofday >=2 and lenofday < 5:
      print("Учимся всего", lenofday, "дня")
    elif lenofday >= 5:
      print("Учимся всего", lenofday, "дней")



    pairs = Soup.select('.study')
    #pprint(pairs)
    #print('df')
    j = 0
    cheat = []
    for i in pairs:
        str = i
        #print(day[j])
       # print(str)

        j+=1
        #print(i.find('span').text)
        t = i.find('span').previous
        tt = t.previous
        ttt = tt.previous;
        cheat.append(ttt.previous)
        cheat.append(t.previous)
        cheat.append(i.find('span').text)
        #print('_________________')

    #print(len(pairs),len(day),len(dayii))

    global d 
    pattern = re.compile('Понедельник')
    columns = Soup.find(text=pattern)
    columns = columns.next
    #columns = columns.next
   # print( columns.next.text)
   # print(columns)
    #mon pairs detecting
    # counter = 0
    # pat = re.compile(group)
    # while True:
    #     columns = columns.next
    #     if isinstance(columns, NavigableString):
    #         continue
    #     # if columns.name != 'span':
    #     #     print('!!!!!!!!!!!!!!!!!!!!!!!!!!!!!')
    #     #     continue
    #     text = columns.text
    #
    #     print(text,'ssds')
    #     if (text.find('Вторник') == -1):
    #
    #         d['mon'].append(columns.text)
    #
    #
    #     else:
    #         break
    #TODO РАССОРТИРОВАТЬ НОРМАЛЬНО
    cases = "Nothing"
    for i in cheat:
       if i == 'Понедельник':
           cases = 'mon'
       if i == 'Вторник':
           cases = 'tue'
       if i == 'Среда':
           cases = 'wed'
       if i == 'Четверг':
           cases = 'thu'
       if i == 'Пятница':
           cases = 'fri'
       if i == 'Суббота':
           cases = 'sat'
       if i == 'Вне сетки расписания':
           cases = 'sun'
       if i == 'Nothin':
           continue
       d[cases].append(i)


# Баг с подсчетом кол-ва дней. Если есть ВНЕ СЕТКИ РАСПИСАНИЯ ПУНКТ - СЧИТАЛ НЕ ВЕРНО СУКА (FIXED)
# Раскрашиваем аутпут
    for i in d['mon']:
      if len(i) == len(group):
        d['mon'].remove(i)
    for i in d['tue']:
      if len(i) == len(group):
        d['tue'].remove(i)
    for i in d['wed']:
      if len(i) == len(group):
        if i != 'Среда':
          d['wed'].remove(i)
    for i in d['thu']:
      if len(i) == len(group):
        d['thu'].remove(i)
    for i in d['fri']:
      if i.find('Группа:') != -1:
        d['fri'].remove(i)
      if len(i) == len(group):
        d['fri'].remove(i)
    for i in d['sat']:
      if i.find('Группа:') != -1:
        d['sat'].remove(i)
      if len(i) == len(group):
        d['sat'].remove(i)


    if debug:
      print('1 parseonline',float(time.time()-t5))




    if debug:
      t6 = time.time()
          
    d['mon'].insert(0,Fore.BLUE+Style.BRIGHT+'Понедельник'+Style.RESET_ALL)
    if len(d['mon']) > 1:
      d['mon'].pop(1)
      for i in range(len(d['mon'])):
       if d['mon'][i].find('пара ') != -1:
         tmp = d['mon'][i]
         d['mon'][i] = ''+Fore.GREEN+Style.BRIGHT+tmp+Style.RESET_ALL
       if d['mon'][i].find('▼') != -1:
        tmp2 = d['mon'][i]
        d['mon'][i] = ''+Fore.BLUE+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
       if d['mon'][i].find('▲') != -1:
        tmp2 = d['mon'][i]
        d['mon'][i] = ''+Fore.RED+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
    else:
      d['mon'].append('Выходной')
    d['tue'].insert(0,Fore.BLUE+Style.BRIGHT+'Вторник'+Style.RESET_ALL)
    if len(d['tue']) > 1:
      d['tue'].pop(1)
      for i in range(len(d['tue'])):
        if d['tue'][i].find('пара ') != -1:
          tmp = d['tue'][i]
          d['tue'][i] = ''+Fore.GREEN+Style.BRIGHT+tmp+Style.RESET_ALL
        if d['tue'][i].find('▼') != -1:
         tmp2 = d['tue'][i]
         d['tue'][i] = ''+Fore.BLUE+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
        if d['tue'][i].find('▲') != -1:
         tmp2 = d['tue'][i]
         d['tue'][i] = ''+Fore.RED+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
    else:
      d['tue'].append('Выходной')
    d['wed'].insert(0,Fore.BLUE+Style.BRIGHT+'Среда'+Style.RESET_ALL)
    if len(d['wed']) > 1:
      d['wed'].pop(1)
      for i in range(len(d['wed'])):
        if d['wed'][i].find('пара ') != -1:
          tmp = d['wed'][i]
          d['wed'][i] = ''+Fore.GREEN+Style.BRIGHT+tmp+Style.RESET_ALL
        if d['wed'][i].find('▼') != -1:
         tmp2 = d['wed'][i]
         d['wed'][i] = ''+Fore.BLUE+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
        if d['wed'][i].find('▲') != -1:
         tmp2 = d['wed'][i]
         d['wed'][i] = ''+Fore.RED+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
    else:
      d['wed'].append('Выходной')
    d['thu'].insert(0,Fore.BLUE+Style.BRIGHT+'Четверг'+Style.RESET_ALL)
    if len(d['thu']) > 1:
      d['thu'].pop(1)
      for i in range(len(d['thu'])):
        if d['thu'][i].find('пара ') != -1:
          tmp = d['thu'][i]
          d['thu'][i] = ''+Fore.GREEN+Style.BRIGHT+tmp+Style.RESET_ALL
        if d['thu'][i].find('▼') != -1:
         tmp2 = d['thu'][i]
         d['thu'][i] = ''+Fore.BLUE+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
        if d['thu'][i].find('▲') != -1:
         tmp2 = d['thu'][i]
         d['thu'][i] = ''+Fore.RED+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
    else:
      d['thu'].append('Выходной')
    d['fri'].insert(0,Fore.BLUE+Style.BRIGHT+'Пятница'+Style.RESET_ALL)
    if len(d['fri']) > 1:
      d['fri'].pop(1)
      for i in range(len(d['fri'])):
        if d['fri'][i].find('пара ') != -1:
          tmp = d['fri'][i]
          d['fri'][i] = ''+Fore.GREEN+Style.BRIGHT+tmp+Style.RESET_ALL
        if d['fri'][i].find('▼') != -1:
         tmp2 = d['fri'][i]
         d['fri'][i] = ''+Fore.BLUE+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
        if d['fri'][i].find('▲') != -1:
         tmp2 = d['fri'][i]
         d['fri'][i] = ''+Fore.RED+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
    else:
      d['fri'].append('Выходной')
    d['sat'].insert(0,Fore.BLUE+Style.BRIGHT+'Суббота'+Style.RESET_ALL)
    if len(d['sat']) > 1:
      d['sat'].pop(1)
      for i in range(len(d['sat'])):
        if d['sat'][i].find('пара ') != -1:
          tmp = d['sat'][i]
          d['sat'][i] = ''+Fore.GREEN+Style.BRIGHT+tmp+Style.RESET_ALL
        if d['sat'][i].find('▼') != -1:
         tmp2 = d['sat'][i]
         d['sat'][i] = ''+Fore.BLUE+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
        if d['sat'][i].find('▲') != -1:
         tmp2 = d['sat'][i]
         d['sat'][i] = ''+Fore.RED+Style.BRIGHT+tmp2[0]+Style.RESET_ALL+tmp2[1:]
    else:
      d['sat'].append('Выходной')
    d['sun'].insert(0,Fore.BLUE+Style.BRIGHT+'Вне сетки расписания'+Style.RESET_ALL)
    if len(d['sun']) > 1:
      d['sun'].pop(1)
      for i in range(len(d['sun'])):
        if d['sun'][i].find('пара ') != -1:
          tmp = d['sun'][i]
          d['sun'][i] = ''+Fore.GREEN+Style.BRIGHT+tmp+Style.RESET_ALL
      d['sun'].pop(1)
    else:
      d['sun'].append('Радуйся если тут пусто')
    global onlinesbor
    onlinesbor = Soup.select('span')[0].text
    try:
      file = open(os.path.expanduser('~/cached/'+group),'r')
      rr = file.read()
      file.close()
      global oflinesbor
      spp = BeautifulSoup(rr,'html.parser')

      oflinesbor = spp.select('span')[0].text
    except IOError:
      print(Fore.RED+Style.BRIGHT+'No cached timetable exists! Try 2 reinstall using pip install -U or u just never cached this group\'s timetable before. If so, do it NOW!'+Style.RESET_ALL)
      oflinesbor = onlinesbor+'FAKE WRONG'

      
      
    if debug:
      print('2 parseonline massive',float(time.time()-t6))



### тут фильтранем неделю

    if nweek != 'c':
        weekz = int(nweek)
    else:
        weekz = week(r)
    if(ns.today):
      weekz = week(r)
      if kek == 7:
        if weekz == 1:
          weekz = 0
        elif weekz == 0:
          weekz = 1
    customizetimetabletomatchcurrentweek(weekz,dz)
    if dz == 'whole':
           customizesimmilar('mon')
           customizesimmilar('tue')
           customizesimmilar('wed')
           customizesimmilar('thu')
           customizesimmilar('fri')
           customizesimmilar('sat')
           customizesimmilar('sun')
           
           prpr(d['mon'])
           print('_________________')
           
           prpr(d['tue'])
           print('_________________')
           
           prpr(d['wed'])
           print('_________________')
           
           prpr(d['thu'])
           print('_________________')
           
           prpr(d['fri'])
           print('_________________')
           
           prpr(d['sat'])
           print('_________________')
           
           prpr(d['sun'])
    elif dz == 'mon':
           customizesimmilar('mon')
           prpr(d['mon'])
    elif dz == 'tue':
           customizesimmilar('tue')
           prpr(d['tue'])
    elif dz == 'wed':
           customizesimmilar('wed')
           prpr(d['wed'])
    elif dz == 'thu':
           customizesimmilar('thu')
           prpr(d['thu'])
    elif dz == 'fri':
           customizesimmilar('fri')
           prpr(d['fri'])
    elif dz == 'sat':
           customizesimmilar('sat')
           prpr(d['sat'])
    elif dz == 'sun':
           customizesimmilar('sun')
           prpr(d['sun'])

def parseofline():
  try:
    file = open(os.path.expanduser('~/cached/'+group),'r')
  except IOError:
    print('File doesnt exists! Try 2 running with --cache key or just ask for timetable once in --online mod')
    exit()
  r = file.read()
  file.close()
  parseonline(r);

def createParser():
  parser = argparse.ArgumentParser(description = 'SUAI Online/Offline timetable')
  return parser
def cachett():
  r = requests.get("http://rasp.guap.ru/").content.decode('utf-8')
  soup = BeautifulSoup(r, "html.parser")
  select = soup.find('option',text=group)
  group_prefix = select.attrs['value']
  r = requests.get("http://rasp.guap.ru/?g="+group_prefix).content.decode('utf-8')
  if not os.path.exists(os.path.expanduser('~/cached')):
    os.makedirs(os.path.expanduser('~/cached'))
  file = open(os.path.expanduser('~/cached/'+group),'w')
  file.write(r)
  file.close()
  tlol = time.time()
  lol = time.localtime(t)
  with open (os.path.expanduser('~/cached/'+group+'.pkl'),'wb') as f:
      pickle.dump(lol,f)

  print('Cached successful! \n File name - ',os.path.abspath(file.name))

def main():
    init()
    parser = createParser()
    parser.add_argument('-o','--online',help=Fore.RED+Style.BRIGHT+'[REQUIRED] Online'+Style.RESET_ALL+' mod',action='store_true')
    parser.add_argument('-f','--offline',help=Fore.RED+Style.BRIGHT+'[REQUIRED] Offline'+Style.RESET_ALL+' mod',action='store_true')
    parser.add_argument('-c','--cache',help=Fore.RED+Style.BRIGHT+'[REQUIRED] Cache '+Style.RESET_ALL+'timetable',action='store_true')
    parser.add_argument('-t','--today',help=Fore.RED+Style.BRIGHT+'Today\'s'+Style.RESET_ALL+' timetable',action='store_true')
    parser.add_argument('-G','--gui',help='Run with GUI[not working]',action='store_true')
    parser.add_argument('-V','--version',help = 'prints version and exit',action='store_true')
    parser.add_argument('-ov','--onlineversion',help = 'prints version stored in online repository and exit',action='store_true')
    parser.add_argument('-v','--verbose',help = 'Verbose output. Debug info.',action='store_true')
    parser.add_argument('-g', '--group', default='5512', help= 'Your group number. DEFAULT = 5512')
    parser.add_argument('-d', '--dz', default='whole', help = 'Day of week. EX.'+Fore.RED+Style.BRIGHT+'TOMORROW(tom)'+Style.RESET_ALL+', MONDAY(mon), TUESDAY(tue), WEDNESDAY(wed) ans so on..')
    parser.add_argument('-w', '--week', default='2', help = 'Red, blue or whole week. EX. 0 = blue ▼, 1 = red ▲, 2 = whole c = current')
    global ns
    ns = parser.parse_args()
    global group
    global nweek
    global zavtra
    global dz
    global t
    global kek
    global debug
    vers = 5.1
    if ns.verbose:
      debug = True
    if(ns.version):
      print('Version:',vers)
      exit()
    if(ns.onlineversion):
      req = requests.get('https://raw.githubusercontent.com/maxan98/grasp/master/setup.py').content.decode('utf-8')
      ver = float(req[784:787])
      print('Online version:',ver)
      exit()
    group = ns.group
    dz = ns.dz
    nweek = ns.week
    zavtra = False
    t = time.time()
    kek = time.localtime(t).tm_wday
    if dz == 'tom':
      zavtra = True
      kek+=1
      ns.today = True
    if(ns.today):
      
      if kek == 0:
        dz = 'mon'
      if kek == 1:
        dz = 'tue'
      if kek == 2:
        dz = 'wed'
      if kek == 3:
        dz = 'thu'
      if kek == 4:
        dz = 'fri'
      if kek == 5:
        dz = 'sat'
      if kek == 6:
        dz = 'sun'
      if kek == 7:
        dz = 'mon'

    if ns.gui:
      root = Tk()
      r = requests.get("http://rasp.guap.ru/").content.decode('utf-8')
      soups = BeautifulSoup(r, "html.parser")
      select = soups.findAll('option')
      values = []
      for i in range (1,len(select)):
          if 'нет' in select[i].text:
              break 
          values.append(select[i].text)
      listgroup = ttk.Combobox(root,height=5,width=8,values = values)
      listgroup.set('5512')
      combod = ['Понедельник','Вторник','Среда','Четверг','Пятница','Суббота','Вне сетки']
      listday = ttk.Combobox(root,height=5,width=8,values = combod)
      listday.set('Понедельник')
      
      
     
     
      onofline = IntVar()
      week = IntVar()
      rbutton1=Radiobutton(root,text='Online',variable=onofline,value=1)
      rbutton2=Radiobutton(root,text='Offline',variable=onofline,value=2)
      rbutton3=Radiobutton(root,text='Red week',variable=week,value=0)
      rbutton4=Radiobutton(root,text='Blue week',variable=week,value=1)
      rbutton5=Radiobutton(root,text='Both weeks',variable=week,value=2)
      button1=Button(root,text='ok',width=25,height=5,bg='black',fg='red',font='arial 14')
      button1.pack(side='bottom')

      rbutton1.pack()
      rbutton2.pack()
      
      listday.pack(side = 'top')
      listgroup.pack(side = 'top')
      rbutton3.pack()
      rbutton4.pack()
      rbutton5.pack()
      root.mainloop()
    if ns.cache:
      cachett()
    elif ns.online:
      parseonline(site())
      #print(onlinesbor)
      if onlinesbor != oflinesbor:
        if query_yes_no('Would you like to cache timetable for this group?'):
          print('Caching..')
          cachett()
      else:
        print('Timetable is up to date')
        req = requests.get('https://raw.githubusercontent.com/maxan98/grasp/master/setup.py').content.decode('utf-8')
        ver = float(req[784:787])
        if vers < ver:
          if query_yes_no('A new version of app is available. Update now?', default = 'no'):
            if subprocess.call(["pip3", "install", "-U","git+https://github.com/maxan98/grasp.git"]) != 0:
              if subprocess.call(["pip", "install", "-U","git+https://github.com/maxan98/grasp.git"]) != 0:
                print('Updated!')
            else:
              print('Updated!')
        else:
          print('App is up to date.')

      
    elif ns.offline:
      

      parseofline()
    else:
      print(Fore.RED+"Wrong usage. Use '-h' for help."+Style.RESET_ALL)
if __name__ == '__main__':
  main()
    